packages/dds/shared-object-base/src/sharedObject.ts

```diff
@@ -247,28 +252,36 @@ export abstract class SharedObjectCore<TEvent extends ISharedObjectEvents = ISha
 	 * {@inheritDoc (ISharedObject:interface).bindToContext}
 	 */
 	public bindToContext(): void {
-		if (this._isBoundToContext) {
-			return;
+		// ensure the method only runs once by removing the implementation
+		// without this the method suffers from re-entrancy issues
+		this.bindToContext = () => {};
```

Comment:
I'm not sure if I should be amazed, terrified, or both 😝 . Isn't it safe right now, because all the code up to setting `_isBoundToContext = true` is sync? So the if check from before would be enough?

Reply:
there is a lot of re-entrancy, even though it is sync, and we can't set the variable util after the binding completes, otherwise things break. This only worked as we fudged the attach state with the services member. This way ensures we only run the function once, and does require adding yet another mutable member to the class.

Reply:
i did debate the safety of this technique, and looked around to see if there were known gotchas, and didn't find any

Reply:
Isn't there a better pattern to run once for this scenario? Most likely there's nothing wrong with this approach but it seems too hacky.

Reply:
you could keep around a boolean that you set at the start of the function and gate on it. but tbh I'm fine with this pattern--it's similar to implementation of `.once()` using `.on()` and immediately unsubscribing and intent is clear.